File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    int preorderIdx = 0;
    Dictionary<int, int> inorderIndexesMap = new();

    public TreeNode BuildTree(int[] preorder, int[] inorder) 
    {
        for(var i = 0; i < inorder.Length; i++)
        {
            inorderIndexesMap[inorder[i]] = i;
        }    
        return BuildSubTree(0, preorder.Length - 1, preorder);
    }

    public TreeNode BuildSubTree(int left, int right, int[] preorder)
    {
        if (left > right)
        {
            return null;
        }
        var currentVal = preorder[preorderIdx++];
        var node = new TreeNode(currentVal);
        var inorderIdx = inorderIndexesMap[currentVal];
        node.left = BuildSubTree(left, inorderIdx - 1, preorder);
        node.right = BuildSubTree(inorderIdx + 1, right, preorder);
        return node;
    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return '#'
        queue = deque([root])
        res = []
        while queue:
            node = queue.popleft()
            if not node:
                res.append('#')
                continue
            res.append(str(node.val))
            queue.append(node.left)
            queue.append(node.right)
        return ','.join(res)
        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        values = data.split(',')
        if not values or values[0] == '#':
            return None
        root = TreeNode(values[0])
        queue = deque([root])
        index = 1

        while queue:
            node = queue.popleft()
            if (values[index] != '#'):
                node.left = TreeNode(int(values[index]))
                queue.append(node.left)
            index += 1
            if (values[index] != '#'):
                node.right = TreeNode(int(values[index]))
                queue.append(node.right)
            index += 1
        return root